[codegen/go] go codegen for explicit providers doesn't compile

This example doesn't compile:https://www.pulumi.com/docs/reference/pkg/aws/s3/bucket/#using-replication-configuration

```
    main.go:7:5: module github.com/pulumi/pulumi-aws/sdk/v3@latest found (v3.32.1), but does not contain package github.com/pulumi/pulumi-aws/sdk/v3/go/aws/providers
    ./main.go:13:13: undefined: providers
    ./main.go:52:22: undefined: aws
    ./main.go:53:22: undefined: aws.Central
```


```go
package main

import (
    "fmt"

    "github.com/pulumi/pulumi-aws/sdk/v3/go/aws/iam"
    "github.com/pulumi/pulumi-aws/sdk/v3/go/aws/providers"
    "github.com/pulumi/pulumi-aws/sdk/v3/go/aws/s3"
    "github.com/pulumi/pulumi/sdk/v2/go/pulumi"
)

func main() {
    pulumi.Run(func(ctx *pulumi.Context) error {
        _, err := providers.Newaws(ctx, "central", &providers.awsArgs{
            Region: pulumi.String("eu-central-1"),
        })
        if err != nil {
            return err
        }
        replicationRole, err := iam.NewRole(ctx, "replicationRole", &iam.RoleArgs{
            AssumeRolePolicy: pulumi.String(fmt.Sprintf("%v%v%v%v%v%v%v%v%v%v%v%v%v", "{\n", "  \"Version\": \"2012-10-17\",\n", "  \"Statement\": [\n", "    {\n", "      \"Action\": \"sts:AssumeRole\",\n", "      \"Principal\": {\n", "        \"Service\": \"s3.amazonaws.com\"\n", "      },\n", "      \"Effect\": \"Allow\",\n", "      \"Sid\": \"\"\n", "    }\n", "  ]\n", "}\n")),
        })
        if err != nil {
            return err
        }
        destination, err := s3.NewBucket(ctx, "destination", &s3.BucketArgs{
            Versioning: &s3.BucketVersioningArgs{
                Enabled: pulumi.Bool(true),
            },
        })
        if err != nil {
            return err
        }
        bucket, err := s3.NewBucket(ctx, "bucket", &s3.BucketArgs{
            Acl: pulumi.String("private"),
            Versioning: &s3.BucketVersioningArgs{
                Enabled: pulumi.Bool(true),
            },
            ReplicationConfiguration: &s3.BucketReplicationConfigurationArgs{
                Role: replicationRole.Arn,
                Rules: s3.BucketReplicationConfigurationRuleArray{
                    &s3.BucketReplicationConfigurationRuleArgs{
                        Id:     pulumi.String("foobar"),
                        Prefix: pulumi.String("foo"),
                        Status: pulumi.String("Enabled"),
                        Destination: &s3.BucketReplicationConfigurationRuleDestinationArgs{
                            Bucket:       destination.Arn,
                            StorageClass: pulumi.String("STANDARD"),
                        },
                    },
                },
            },
        }, pulumi.Provider(aws.Central))
        if err != nil {
            return err
        }
        replicationPolicy, err := iam.NewPolicy(ctx, "replicationPolicy", &iam.PolicyArgs{
            Policy: pulumi.All(bucket.Arn, bucket.Arn, destination.Arn).ApplyT(func(_args []interface{}) (string, error) {
                bucketArn := _args[0].(string)
                bucketArn1 := _args[1].(string)
                destinationArn := _args[2].(string)
                return fmt.Sprintf("%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v%v", "{\n", "  \"Version\": \"2012-10-17\",\n", "  \"Statement\": [\n", "    {\n", "      \"Action\": [\n", "        \"s3:GetReplicationConfiguration\",\n", "        \"s3:ListBucket\"\n", "      ],\n", "      \"Effect\": \"Allow\",\n", "      \"Resource\": [\n", "        \"", bucketArn, "\"\n", "      ]\n", "    },\n", "    {\n", "      \"Action\": [\n", "        \"s3:GetObjectVersion\",\n", "        \"s3:GetObjectVersionAcl\"\n", "      ],\n", "      \"Effect\": \"Allow\",\n", "      \"Resource\": [\n", "        \"", bucketArn1, "/*\"\n", "      ]\n", "    },\n", "    {\n", "      \"Action\": [\n", "        \"s3:ReplicateObject\",\n", "        \"s3:ReplicateDelete\"\n", "      ],\n", "      \"Effect\": \"Allow\",\n", "      \"Resource\": \"", destinationArn, "/*\"\n", "    }\n", "  ]\n", "}\n"), nil
            }).(pulumi.StringOutput),
        })
        if err != nil {
            return err
        }
        _, err = iam.NewRolePolicyAttachment(ctx, "replicationRolePolicyAttachment", &iam.RolePolicyAttachmentArgs{
            Role:      replicationRole.Name,
            PolicyArn: replicationPolicy.Arn,
        })
        if err != nil {
            return err
        }
        return nil
    })
}
```

--- a/CHANGELOG_PENDING.md
+++ b/CHANGELOG_PENDING.md
@@ -7,3 +7,7 @@
 
 - [cli] Only log github request headers at log level 11.
   [#10140](https://github.com/pulumi/pulumi/pull/10140)
+
+- [codegen/go] Support program generation, `pulumi convert` for programs that create explicit
+  provider resources.
+  [#10132](https://github.com/pulumi/pulumi/issues/10132)

--- a/pkg/codegen/go/gen_program.go
+++ b/pkg/codegen/go/gen_program.go
@@ -327,12 +327,16 @@ func (g *generator) collectTypeImports(program *pcl.Program, t schema.Type, impo
 	}
 
 	var tokenRange hcl.Range
-	pkg, mod, _, _ := pcl.DecomposeToken(token, tokenRange)
-	vPath, err := g.getVersionPath(program, pkg)
+	pkg, mod, name, _ := pcl.DecomposeToken(token, tokenRange)
+	pkgName := pkg
+	if pkg == "pulumi" && mod == "providers" {
+		pkgName = name
+	}
+	vPath, err := g.getVersionPath(program, pkgName)
 	if err != nil {
 		panic(err)
 	}
-	imports.Add(g.getPulumiImport(pkg, vPath, mod))
+	imports.Add(g.getPulumiImport(pkg, vPath, mod, name))
 }
 
 // collect Imports returns two sets of packages imported by the program, std lib packages and pulumi packages
@@ -345,16 +349,17 @@ func (g *generator) collectImports(
 	for _, n := range program.Nodes {
 		if r, isResource := n.(*pcl.Resource); isResource {
 			pkg, mod, name, _ := r.DecomposeToken()
-			if pkg == "pulumi" && mod == "providers" {
-				pkg = name
-			}
 
-			vPath, err := g.getVersionPath(program, pkg)
+			pkgName := pkg
+			if pkg == "pulumi" && mod == "providers" {
+				pkgName = name
+			}
+			vPath, err := g.getVersionPath(program, pkgName)
 			if err != nil {
 				panic(err)
 			}
 
-			pulumiImports.Add(g.getPulumiImport(pkg, vPath, mod))
+			pulumiImports.Add(g.getPulumiImport(pkg, vPath, mod, name))
 		}
 		if _, isConfigVar := n.(*pcl.ConfigVariable); isConfigVar {
 			pulumiImports.Add("\"github.com/pulumi/pulumi/sdk/v3/go/pulumi/config\"")
@@ -366,7 +371,7 @@ func (g *generator) collectImports(
 					tokenArg := call.Args[0]
 					token := tokenArg.(*model.TemplateExpression).Parts[0].(*model.LiteralValueExpression).Value.AsString()
 					tokenRange := tokenArg.SyntaxNode().Range()
-					pkg, mod, _, diagnostics := pcl.DecomposeToken(token, tokenRange)
+					pkg, mod, name, diagnostics := pcl.DecomposeToken(token, tokenRange)
 
 					contract.Assert(len(diagnostics) == 0)
 
@@ -374,7 +379,7 @@ func (g *generator) collectImports(
 					if err != nil {
 						panic(err)
 					}
-					pulumiImports.Add(g.getPulumiImport(pkg, vPath, mod))
+					pulumiImports.Add(g.getPulumiImport(pkg, vPath, mod, name))
 				} else if call.Name == pcl.IntrinsicConvert {
 					g.collectConvertImports(program, call, pulumiImports)
 				}
@@ -463,10 +468,13 @@ func (g *generator) getGoPackageInfo(pkg string) (GoPackageInfo, bool) {
 	return info, ok
 }
 
-func (g *generator) getPulumiImport(pkg, vPath, mod string) string {
+func (g *generator) getPulumiImport(pkg, vPath, mod, name string) string {
 	info, _ := g.getGoPackageInfo(pkg)
 	if m, ok := info.ModuleToPackage[mod]; ok {
 		mod = m
+	} else if pkg == "pulumi" && mod == "providers" {
+		pkg = name
+		mod = ""
 	}
 
 	imp := fmt.Sprintf("github.com/pulumi/pulumi-%s/sdk%s/go/%s/%s", pkg, vPath, pkg, mod)
@@ -608,7 +616,13 @@ func (g *generator) genResource(w io.Writer, r *pcl.Resource) {
 		g.genTemps(w, temps)
 	}
 
-	modOrAlias := g.getModOrAlias(pkg, mod)
+	var modOrAlias string
+	if pkg == "pulumi" && mod == "providers" {
+		modOrAlias = typ
+		typ = "Provider"
+	} else {
+		modOrAlias = g.getModOrAlias(pkg, mod)
+	}
 
 	instantiate := func(varName, resourceName string, w io.Writer) {
 		if g.scopeTraversalRoots.Has(varName) || strings.HasPrefix(varName, "__") {

--- a/pkg/codegen/pcl/binder_resource.go
+++ b/pkg/codegen/pcl/binder_resource.go
@@ -64,26 +64,26 @@ func (b *binder) bindResourceTypes(node *Resource) hcl.Diagnostics {
 		return hcl.Diagnostics{unknownPackage(pkg, tokenRange)}
 	}
 
+	var res *schema.Resource
 	var inputProperties, properties []*schema.Property
-	if !isProvider {
-		var res *schema.Resource
-		if r, tk, ok, err := pkgSchema.LookupResource(token); err != nil {
-			return hcl.Diagnostics{resourceLoadError(token, err, tokenRange)}
-		} else if !ok {
-			return hcl.Diagnostics{unknownResourceType(token, tokenRange)}
-		} else {
-			res = r
-			token = tk
-		}
-		node.Schema = res
-		inputProperties, properties = res.InputProperties, res.Properties
-	} else {
-		config, err := pkgSchema.schema.Config()
+	if isProvider {
+		r, err := pkgSchema.schema.Provider()
 		if err != nil {
 			return hcl.Diagnostics{resourceLoadError(token, err, tokenRange)}
 		}
-		inputProperties, properties = config, config
+		res = r
+	} else {
+		r, tk, ok, err := pkgSchema.LookupResource(token)
+		if err != nil {
+			return hcl.Diagnostics{resourceLoadError(token, err, tokenRange)}
+		} else if !ok {
+			return hcl.Diagnostics{unknownResourceType(token, tokenRange)}
+		}
+		res = r
+		token = tk
 	}
+	node.Schema = res
+	inputProperties, properties = res.InputProperties, res.Properties
 	node.Token = token
 
 	// Create input and output types for the schema.

--- a/pkg/codegen/testing/test/program_driver.go
+++ b/pkg/codegen/testing/test/program_driver.go
@@ -132,8 +132,6 @@ var PulumiPulumiProgramTests = []ProgramTest{
 	{
 		Directory:   "aws-resource-options",
 		Description: "Resource Options",
-		SkipCompile: codegen.NewStringSet("go"),
-		// Blocked on go: TODO[pulumi/pulumi#8076]
 	},
 	{
 		Directory:   "aws-secret",

--- a/pkg/codegen/testing/test/testdata/aws-resource-options-pp/go/aws-resource-options.go
+++ b/pkg/codegen/testing/test/testdata/aws-resource-options-pp/go/aws-resource-options.go
@@ -1,15 +1,15 @@
 package main
 
 import (
-	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws/providers"
+	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws"
 	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws/s3"
 	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
 )
 
 func main() {
 	pulumi.Run(func(ctx *pulumi.Context) error {
-		provider, err := providers.Newaws(ctx, "provider", &providers.awsArgs{
-			Region: "us-west-2",
+		provider, err := aws.NewProvider(ctx, "provider", &aws.ProviderArgs{
+			Region: pulumi.String("us-west-2"),
 		})
 		if err != nil {
 			return err

